boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,293 @@
+
+from boat_simulator.common.types import Scalar
+from boat_simulator.common.utils import bound_to_180
+
+
+class ActuatorController(ABC):
+    """Abstract class for rudder and sail actuation mechanisms.
+
+    Attributes:
+        `current_heading` (Scalar): Current boat heading direction in degrees.
+        `desired_heading` (Scalar): Target boating heading direction in degrees.
+        `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+        `time_step` (Scalar): Time per iteration given in degrees/second.
+        `kp` (Scalar): Proportional constant when calculating error.
+        `cp` (Scalar): Timeseries of error values computed over time.
+        `prev_control` (list): Timeseries of control angles computed over time.
+        `time_series` (list): Timeseries of running time values corresponding to control angles
+
+    """
+
+    def __init__(
+        self,
+        current_heading: Scalar,
+        desired_heading: Scalar,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        kp: Scalar,
+        cp: Scalar,
+        max_angle_range: tuple,
+    ):
+        """Initializes the class attributes. Note that this class cannot be directly instantiated.
+
+        Args:
+            `current_heading` (Scalar): Current boat heading direction in degrees.
+            `desired_heading` (Scalar): Target boating heading direction in degrees.
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in degrees/second.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Timeseries of error values computed over time.
+            `prev_control` (list): Timeseries of control angles computed over time.
+            `time_series` (list): Timeseries of running time values related to control angles.
+            `max_angle_range` (tuple): Tuple of minimum to maximum control angle range
+        """
+        self.current_heading = current_heading
+        self.desired_heading = desired_heading
+        self.current_control_ang = current_control_ang
+        self.time_step = time_step
+        self.kp = kp
+        self.cp = cp
+        self.max_angle_range = max_angle_range
+        self.prev_control: List[Scalar] = list()
+        self.time_series: List[Scalar] = list()
+
+        self.time_series.append[0]  # append initial starting time 0
+        self.prev_control.append[self.current_control_ang]  # append starting control angle
+
+    def append_heading(self, past_heading: Scalar) -> None:
+        """Appends the previous heading to heading list and updates the current
+        time taken to iterate to the current rudder position
+
+        Args:
+            `past_heading` (Scalar): The previous heading to be added
+
+        """
+        self.prev_control.append(past_heading)
+        self.time_series.append(self.time_series[-1] + self.time_step)
```

Comment:
Do you need to keep track of this?

Reply:
Not needed to keep track for implementation as well haha.